### PR TITLE
Handle AssetIds with invalid URI characters

### LINKED ISSUE #1134: Watcher crashed for emacs temp file
Edited a file in emacs, which created a temporary file, while running `build_runner serve`, and got the crash listed below:
```
FormatException: Invalid character (at character 37)
examples.registration_form/address/.#address_component.html
                                    ^

#0      _Uri._fail (dart:core/uri.dart:1597)
#1      _Uri._normalize (dart:core/uri.dart:2292)
#2      _Uri._normalizeOrSubstring (dart:core/uri.dart:2248)
#3      _Uri._makePath (dart:core/uri.dart:2096)
#4      new _Uri (dart:core/uri.dart:1497)
#5      AssetId.uri (package:build/src/asset/id.dart:102:13)
#6      PackageGraphWatcher._watch.<anonymous closure>.<anonymous closure> (package:build_runner/src/watcher/graph_watcher.dart:84:54)
#7      _RootZone.runUnaryGuarded (dart:async/zone.dart:1316)
#8      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:330)
#9      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:257)
#10     _ForwardingStreamSubscription._add (dart:async/stream_pipe.dart:132)
#11     _MapStream._handleData (dart:async/stream_pipe.dart:232)
#12     _ForwardingStreamSubscription._handleData (dart:async/stream_pipe.dart:164)
#13     _RootZone.runUnaryGuarded (dart:async/zone.dart:1316)
#14     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:330)
#15     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:257)
#16     _SyncBroadcastStreamController._sendData (dart:async/broadcast_stream_controller.dart:379)
#17     _BroadcastStreamController.add (dart:async/broadcast_stream_controller.dart:254)
#18     _RootZone.runUnaryGuarded (dart:async/zone.dart:1316)
#19     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:330)
#20     _DelayedData.perform (dart:async/stream_impl.dart:578)
#21     _StreamImplEvents.handleNext (dart:async/stream_impl.dart:694)
#22     _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:654)
#23     _microtaskLoop (dart:async/schedule_microtask.dart:41)
#24     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
#25     _Timer._runTimers (dart:isolate-patch/dart:isolate/timer_impl.dart:391)
#26     _Timer._handleMessage (dart:isolate-patch/dart:isolate/timer_impl.dart:416)
#27     _RawReceivePortImpl._handleMessage (dart:isolate-patch/dart:isolate/isolate_patch.dart:165)
```

### PR DESCRIPTION
**Pull request by natebosch**
Fixes #1134

Rewrite the URI encoding for AssetId to be more robust by using the
built in URI escaping when using `pathSegments` instead of `path`.

Delete a duplicate of the AssetId test file.

### CODE REVIEW

**Comment by natebosch** [comment]
We had talked about just throwing away any file path that had an invalid character, but I think this is probably the better fix.

File: build_runner/test/asset/id_test.dart [hunk 1]
```diff
@@ -1,71 +0,0 @@
-// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
```
Comment (jakemac53) [inline build_runner/test/asset/id_test.dart]: were the rest of these all dupes?

File: build/lib/src/asset/id.dart [hunk 2]
```diff
@@ -154,3 +151,12 @@ String _normalizePath(String path) {
...
   return p.posix.normalize(path);
 }
+
+Uri _constructUri(AssetId id) {
+  final originalSegments = p.split(id.path);
+  final isLib = originalSegments.first == 'lib';
+  final scheme = isLib ? 'package' : 'asset';
+  final pathSegments = isLib ? originalSegments.skip(1) : originalSegments;
+  return new Uri(
+      scheme: scheme, pathSegments: [id.package]..addAll(pathSegments));
```
Comment (jakemac53) [inline build/lib/src/asset/id.dart]: so `pathSegments` escapes and `path` doesn't?

Comment (natebosch) [reply build/lib/src/asset/id.dart]: `path` escapes some stuff, but not everything. It leaves delimiters, even when they are invalid.
https://github.com/dart-lang/sdk/blob/80a571d5460bbc734e709d7437af2c7df11e582a/sdk/lib/core/uri.dart#L85

`pathSegments` escapes everything which is, I think, what we want.

File: build_runner/test/asset/id_test.dart (hunk 1 above)
Comment (natebosch) [reply build_runner/test/asset/id_test.dart]: no, the other ones in this directory are testing build_runner stuff.

**Pull request merged**